File: cpp/one/a053.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int n = nums.size();
        vector<int> sums(n, 0);
        sums[0] = nums[0];
        
        int mx = sums[0];
        for (int i = 1; i < n; i++) {
            sums[i] = max (sums[i-1] + nums[i], nums[i]);
            mx = max (mx, sums[i]);
        }
        return mx;
    }
};

File: python/a053.py
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 0:
            return 0
        
        dp = [0 for _ in range(len(nums))]  # DP table, store subarray sum
        maxSub = dp[0] = nums[0]
        
        for i in range(1, len(nums)):
            # Disregard previous sum if less than 0
            dp[i] = (dp[i-1] + nums[i])  if dp[i-1] >= 0 else nums[i]
            maxSub = max(maxSub, dp[i])
        return maxSub